### DISCUSSION TITLE
Email address as SAN in template

### DISCUSSION THREAD

**Discussion by jwh** [smallstep/certificates#983] [Q&A]
Hi guys,

I'm trying to write a template that takes the email address and CN and adds SANs for those, the CN works (as dns:<CN>) however I can't figure out how to take the email address from the subject and add use that as san email:<email> - probably being a potato but is this possible? Checked x509util et al but don't see how I can access multistring in a template etc

thanks

**Comment by jwh** [comment]
Any ideas? I really can't figure this out from the documentation or the code, and even valid looking templates result in errors like ``error="error applying certificate template: invalid character '[' looking for beginning of object key string at offset 138"`` which aren't helpful at all really.

**Reply by maraino** [reply]
A `MultiString` is basically `[]string`, but it also allows a string as a parameter, meaning that these two configurations will be equivalent:
* "emailAddresses": ["my@email.com"]
* "emailAddresses": "my@email.com"

I don't know where you are trying to get the email address, but let's assume it is coming from a token with an "email" claim, and you're not using the OIDC provisioner that already manages that case. This template will set the default SANs of the provisioner you are using and the email available in the token:

```go
{
	"subject": {{ toJson .Subject }},
	"sans": {{ toJson .SANs }},
{{- if .Token.email }}
	"emailAddresses": {{ toJson .Token.email }}
{{- end }}
{{- if typeIs "*rsa.PublicKey" .Insecure.CR.PublicKey }}
	"keyUsage": ["keyEncipherment", "digitalSignature"],
{{- else }}
	"keyUsage": ["digitalSignature"],
{{- end }}
	"extKeyUsage": ["serverAuth", "clientAuth"]
}
```

That template is the default one with the addition of 
```go
{{- if .Token.email }}
	"emailAddresses": {{ toJson .Token.email }}
{{- end }}
```

Meaning that if the token has an email claim (`"email": "my@email.com"`), we will add it, and the rendered template will have:
```
	"emailAddresses": "my@email.com"
```

Here we're using the simple string format of `MultiString`, and we're using `toJson` to encode the email properly. At the same time, if `.Token.email` is, for example `["my@email1.com", "my@email2.com"]` it will add both emails.
